modules/core/04-channel/keeper/upgrade.go

```diff
@@ -86,17 +88,249 @@ func (k Keeper) WriteUpgradeInitChannel(
+			return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
+		}
+		return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, "upgrade timed out")
+	}
+
+	switch channel.State {
+	case types.OPEN:
+		upgradeSequence = uint64(0)
+		if seq, found := k.GetUpgradeSequence(ctx, portID, channelID); found {
+			upgradeSequence = seq
+		}
+
+		// if the counterparty upgrade sequence is ahead then fast forward so both channel ends are using the same sequence for the current upgrade
+		if counterpartyUpgradeSequence > upgradeSequence {
+			upgradeSequence = counterpartyUpgradeSequence
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+		} else {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			// the upgrade sequence is incremented so both sides start the next upgrade with a fresh sequence.
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+
+			// TODO: emit error receipt events
+
+			// do we want to return upgrade sequence here to include in response??
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", errorReceipt.GetSequence())
+		}
+
+		// this is first message in upgrade handshake on this chain so we must store original channel in restore channel path
+		// in case we need to restore channel later.
+		k.SetUpgradeRestoreChannel(ctx, portID, channelID, channel)
+	case types.INITUPGRADE:
+		upgradeSequence, found = k.GetUpgradeSequence(ctx, portID, channelID)
+		if !found {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		if upgradeSequence != counterpartyUpgradeSequence {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not equal to the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+
+			// set to the max of the two
+			if counterpartyUpgradeSequence > upgradeSequence {
+				upgradeSequence = counterpartyUpgradeSequence
+				k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+			}
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
```

Comment:
The upgrade sequence mentioned in this error message could be different than the one in the error receipt if it gets updated in the if clause just above.